sycl/include/CL/sycl/intel/builtins.hpp

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define CONSTANT_AS __attribute__((opencl_constant))
+// Note: __format string is declared in constant address space to be compatible
+// with OpenCL C
+extern int __spirv_ocl_printf(const CONSTANT_AS char *__format, ...);
+#else
+#define CONSTANT_AS
+#endif
+
+namespace cl {
+namespace sycl {
+namespace intel {
+namespace experimental {
+
+// Provides functionality to print data from kernels in a C way:
+// - On non-host devices this function is directly mapped to printf from
+//   OpenCL C
+// - On host device, this function should be equivalent to standard printf
+//   function from C/C++.
+//
+// Please refer to corresponding section in OpenCL C specification to find
+// information about format string and its differences from standard C rules.
+//
+// This function is placed under 'experimental' namespace on purpose, because it
+// has too much caveats you need to be aware of before using it. Please find
+// them below and read carefully before using it:
+//
+// - According to the OpenCL spec, the format is must be
+// resolvable at compile time i.e. cannot be dynamically created by the
+// executing program, itself.
+//
+// - According to the OpenCL spec, the format string must reside in constant
+// address space. This requires to perform "tricky" declarations of them, see
+// test/built-ins/printf.cpp for examples
+// FIXME: this potentially can be done on SYCL FE side automatically
+//
+// - The format string is interpreted according to the OpenCL C spec, where all
+// data types has fixed bitwidth, opposed to C++ types which doesn't guarantee
+// the exact width of particular data types (except, may be, char). This might
+// lead to unexpected result, for example: %ld in OpenCL C means that printed
+// argument has 'long' type which is 64-bit wide by the OpenCL C spec. However,
+// by C++ spec long is just at least 32-bit wide, so, you need to ensure (by
+// performing a cast, for example) that if you use %ld specifier, you pass
+// 64-bit argument to the cl::sycl::experimental::printf
+//
+// - OpenCL spec defines several additional features, like, for example, 'v'
+// modifier which allows to print OpenCL vectors: note that these features are
+// not available on host device and therefore their usage should be either
+// guarded using __SYCL_DEVICE_ONLY__ preprocessor macro or avoided in favor
+// of more portable solutions if needed
+//
+// Additionally, there are several interesting features which are exposed by
+// this printf function:
+//
+// - You can print vector data types using 'v' modifier:
+//   #ifdef __SYCL_DEVICE_ONLY__ // this is only available on real devices
+//     auto vec = cl::sycl::vec<int, 8> v(1, 2, 3, 4, 5, 6, 7, 8);
+//     using ocl_vec_type = cl::sycl::vec<int, 8>::vector_t;
```

Comment:
And if I have `cl::sycl::vec<long, 8>` on Windows, I have to use `%v8d` instead of `%v8ld`, because `long` is 32bit there, right? We definitely need a correct format diagnostic here (not with this patch). Or maybe Clang can adjust a format string according to the OpenCL rules, since it is guaranteed to be a compile-time constant.